thunder/distributed/transforms/fsdp_v2.py

```diff
@@ -153,92 +153,91 @@ def transform_module(
-                    )
-                for pn, _ in submodule.named_buffers(recurse=False, prefix=module_name):
-                    tdist.broadcast(
-                        thunder_model.get_buffer(pn), src=self.broadcast_from, group=self.process_group, async_op=False
-                    )
-
-            for pn, p in submodule.named_parameters(recurse=False, prefix=module_name):
-                # If there are shared params in the original user Module, we reuse the sharded copy created from the original parameter below.
-                # This way we re-create parameter sharing in thunder's copy of the Module.
-                if p in shared_params:
-                    # Shared param names : current param - base param
-                    self.shared_params_name[pn] = shared_params[p]["param_name"]
-                    # Re-use the previous copy of this parameter.
-                    thunder_model._overrides_parameters[pn] = shared_params[p]["param_copy"]
-                    self.sharded_params[pn] = shared_params[p]["param_shard_meta"]
-                    continue
-
-                # if we don't have an override or it is just the original, do create a copy
-                if thunder_model._overrides_parameters.get(pn, p) is p:
-                    thunder_model._overrides_parameters[pn] = copy.copy(p)
-                # we collect shapes and devices because we do not know if other transforms also change it...
-                old_shape = thunder_model._overrides_parameters[pn].shape
-                _shard_param(
-                    thunder_model._overrides_parameters[pn], global_rank, world_size, pn, allow_padding_for_fsdp=True
+        # Note that .named_parameters / .named_buffers used below only return a duplicated parameter once.
+
+        shared_names = thunder_model._get_shared_names()
+        self.shared_params_name = {}
+
+        # For materialized parameters and buffers, we move them to the target device as necessary
+        # for un-materialized parameters and buffers, we set the ._thunder_device
+
+        is_fully_materialized = True
+        for n, p in thunder_model.named_parameters():
+            for nn in shared_names[n]:
+                if nn != n:
+                    self.shared_params_name[nn] = n
+            try:
+                orig_p = thunder_model._model.get_parameter(n)
+            except AttributeError as _:
+                orig_p = None
+            if p.is_meta:
+                is_fully_materialized = False
+                p._thunder_device = self.device
+                if orig_p is not None:
+                    orig_p._thunder_device = self.device
+                # TODO: check if devic_adjustments are still needed
+                for nn in shared_names[n]:
+                    device_adjustments[nn] = self.device
+            elif p.device != self.device:
+                new_p = torch.nn.Parameter(p.to(device=self.device), requires_grad=p.requires_grad)
```

Comment:
wouldn't this line want to be in the context of `torch.no_grad()`?

```suggestion
                new_p = torch.nn.Parameter(p.to(device=self.device), requires_grad=p.requires_grad)
```

Reply:
```suggestion
                with torch.no_grad():
                    new_p = torch.nn.Parameter(p.to(device=self.device), requires_grad=p.requires_grad)
```